### PR TITLE
Add `max_transfer_size/0`

### PR DESCRIPTION
**Pull request by cocoa-xu**
Linux put the maximum buffer size of the SPI device at `/sys/module/spidev/parameters/bufsiz`. 

The amount of data in a single transfer must be less than or equal to this number, otherwise, `ioctl` will return a negative value indicating errors. Subsequently, users will get `{:error, :transfer_failed}` in Erlang/Elixir.

This PR added a `max_transfer_size` function to retrieve the maximum buffer size. If `/sys/module/spidev/parameters/bufsiz` is not available, then it returns ~`:unknown`~ `4096`, a safe minimum size in Linux.

### CODE REVIEW

File: lib/spi.ex [hunk 1]
```diff
@@ -88,6 +88,14 @@ defmodule Circuits.SPI do
   @spec info() :: map()
   defdelegate info(), to: Nif
 
+  @doc """
+  Return max buffer size of the low level SPI interface
+
+  This may be helpful when deciding transfer size.
+  """
+  @spec max_buf_size() :: integer() | :unknown
+  defdelegate max_buf_size(), to: Nif
```
Comment (fhunleth) [inline lib/spi.ex:108]:
```suggestion
  @spec max_transfer_size() :: integer()
  defdelegate max_transfer_size(), to: Nif
```

In addition to the rename, I removed `:unknown`. I think we can spare users having to handle `:unknown` by giving them a minimum supported size.

File: src/hal_spidev.c [hunk 2]
```diff
@@ -32,6 +34,39 @@ ERL_NIF_TERM hal_info(ErlNifEnv *env)
     return info;
 }
 
+ERL_NIF_TERM hal_max_buf_size(ErlNifEnv *env)
+{
+    static int cached = 0;
+    static ERL_NIF_TERM max_buf_size;
```
Comment (fhunleth) [inline src/hal_spidev.c:40]: There's a highly improbable race condition between the setting of `cached` and `max_buf_size` since `cached` is set after `max_buf_size` and there's no mutex. I don't think it matters, though. Could you remove the cache and just read it every time?

Since this is Linux `spidev`, a 4096 minimum size is safe it it can't be read. That's the default and anyone setting it lower should know what they're doing.

File: src/hal_spidev.c [hunk 3]
```diff
@@ -32,6 +34,39 @@ ERL_NIF_TERM hal_info(ErlNifEnv *env)
...
+    if (cached != 0) {
+        return max_buf_size;
+    }
+
+    // Linux put this information (if available) in /sys/module/spidev/parameters/bufsiz
+    FILE *file = fopen("/sys/module/spidev/parameters/bufsiz","r");
+    if (file != NULL) {
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-result"
+        fscanf(file, "%"PRIu64, &bufsiz);
```
Comment (fhunleth) [inline src/hal_spidev.c:53]: Check the error here rather than ignore it.

File: src/hal_spidev.c [hunk 4]
```diff
@@ -78,7 +113,8 @@ int hal_spi_transfer(int fd,
                      const struct SpiConfig *config,
                      const uint8_t *to_write,
                      uint8_t *to_read,
-                     size_t len)
+                     size_t len,
+                     size_t chunk_size)
```
Comment (fhunleth) [inline src/hal_spidev.c:117]: Let's discuss chunking in a separate issue. My initial thought is this should be kept in Elixir since there's probably not much of a performance improvement (or I'd hope not), and it reduces the max time spent in the NIF.

**Review by fhunleth — COMMENTED** [review]
This looks good. I'm seeing changes come in and now I'm not sure sure how it stands, so I'm going to post the initial comments. The max transfer size makes sense. How about we get that one in first.

**Comment by cocoa-xu** [comment]
Added `transfer/3` for performance.

We have to split data into proper chunks manually in Erlang/Elixir before this.

```elixir
for xfdata <-
          bytes_to_be_sent
          |> Enum.chunk_every(chunk_size)
          |> Enum.map(&Enum.into(&1, <<>>, fn bit -> <<bit::8>> end)) do
      {:ok, _ret} = Circuits.SPI.transfer(spi, xfdata)
    end
```

With `transfer/3` we can use the following code

```elixir
{:ok, _ret} = Circuits.SPI.transfer(spi, xfdata, chunk_size)
```

In an ideal world, there should be no performance difference when you split the bytes in Erlang/Elixir or C. But we do have overheads here...

I observed ~22% performance on a Raspberry Pi 4 (SPI@125MHz) improvement after switching to `transfer/3`.

File: lib/spi.ex (hunk 1 above)
Comment (cocoa-xu) [reply lib/spi.ex]: I agree with you. I was thinking about this, but I didn't know the minimal safe size. But it's definitely better to have some default value instead of doing more handling.

File: src/hal_spidev.c (hunk 4 above)
Comment (cocoa-xu) [reply src/hal_spidev.c:109]: Yeah, they really should be quite close, but my experiments found that when transferring 2457600 bits @ 125MHz on a Raspberry Pi 4, there was ~22% performance improvement (`0.33` vs. `0.43` seconds) if we do this in C. Maybe we can set up some tests to further prove this?

**Comment by fhunleth** [comment]
Could you remove the chunking code for another PR? Erlang has better ways of splitting binaries and reassembling using iodata routines is more efficient. Plus the iodata can be pushed deeper.

File: lib/spi.ex [hunk 5]
```diff
@@ -88,6 +88,14 @@ defmodule Circuits.SPI do
   @spec info() :: map()
   defdelegate info(), to: Nif
 
+  @doc """
+  Return max buffer size of the low level SPI interface
+
+  This may be helpful when deciding transfer size.
```
[the remark this answers is not in the document]
Comment (cocoa-xu) [reply lib/spi.ex]: > Could you remove the chunking code for another PR? Erlang has better ways of splitting binaries and reassembling using iodata routines is more efficient. Plus the iodata can be pushed deeper.

Sure! Hang on a second :)

**Comment by cocoa-xu** [comment]
> Could you remove the chunking code for another PR? Erlang has better ways of splitting binaries and reassembling using iodata routines is more efficient. Plus the iodata can be pushed deeper.

Sure!

File: src/hal_spidev.c [hunk 6]
```diff
@@ -32,6 +34,31 @@ ERL_NIF_TERM hal_info(ErlNifEnv *env)
...
+
+    if (bufsiz == 0) {
+        // if /sys/module/spidev/parameters/bufsiz is not available
+        // then we return 4096, a safe minimum size
+        max_buf_size = enif_make_uint64(env, 4096);
+    } else {
+        max_buf_size = enif_make_uint64(env, bufsiz);
+    }
+
+    return max_buf_size;
```
Comment (fhunleth) [inline src/hal_spidev.c:59]:
```suggestion
    if (bufsiz == 0) {
        // if /sys/module/spidev/parameters/bufsiz is not available
        // then we return 4096, a safe minimum size
        buf_size = 4096;
    } 
    
    return enif_make_uint64(env, bufsiz);
```

File: lib/spi.ex [hunk 7]
```diff
@@ -88,6 +88,21 @@ defmodule Circuits.SPI do
...
+  may be capped by the low level SPI interface. For example,
+  the Linux `spidev` driver allocates its transfer buffer at
+  initialization based on the `bufsiz` parameter and rejects
+  requests that won't fit.
+
+  If you're sending large amounts of data over SPI, use this
+  function to determine how to split up large messages.
+  """
+  @spec max_buf_size() :: non_neg_integer()
+  defdelegate max_buf_size(), to: Nif
```
Comment (fhunleth) [inline lib/spi.ex:104]: How about this rename? 

```suggestion
  @spec max_transfer_size() :: non_neg_integer()
  defdelegate max_transfer_size(), to: Nif
```

File: src/hal_stub.c [hunk 8]
```diff
@@ -26,6 +27,13 @@ ERL_NIF_TERM hal_info(ErlNifEnv *env)
     return info;
 }
 
+ERL_NIF_TERM hal_max_buf_size(ErlNifEnv *env)
+{
+    uint64_t max = 0;
+    max -= 1;
+    return enif_make_uint64(env, max);
```
Comment (fhunleth) [inline src/hal_stub.c:34]:
```suggestion
    // Use Linux's default maximum transfer size
    return enif_make_uint64(env, 4096);
```

File: src/hal_spidev.c (hunk 6 above)
Comment (cocoa-xu) [reply src/hal_spidev.c]: This is why I shouldn't code at 3 am lol

**Comment by cocoa-xu** [comment]
So, I ran some benchmarks and here is the [result](https://gist.github.com/cocoa-xu/0919081bb954953aa9d6f9aa934cb47c).

I guess there are some reasons that `Enum.chunk_every` does not support `BitString`. We have to convert `BitString` to `List`, pass the list to `Enum.chunk_every` and then map each element back to `BitString`, which consumes a lot of time. I don't know if there are any better approaches based on this.

```elixir
def to_u8_chunk_every_map(binary, chunk_size) do
    binary
    |> :binary.bin_to_list()
    |> Enum.chunk_every(chunk_size)
    |> Enum.map(&IO.iodata_to_binary(&1))
end
```

But when we do it manually, it's basically instantaneous.

```elixir
def manually(binary, chunk_size) do
    total_bytes = byte_size(binary)
    full_chunks = div(total_bytes, chunk_size)
    chunks = 
        if full_chunks > 0 do
           for i <- 0..(full_chunks-1), reduce: [] do
                acc -> [:binary.part(binary, chunk_size * i, chunk_size) | acc]
            end
        else
            []  
        end
    remaining = rem(total_bytes, chunk_size)
    chunks = if remaining > 0 do
            [:binary.part(binary, chunk_size * full_chunks, remaining) | chunks]
        else
            chunks
        end
    Enum.reverse(chunks)
end
```

In conclusion, maybe we don't really need `transfer/3`.

**Review by fhunleth — APPROVED** [review]
This looks great. Thanks for the revisions and the benchmarks.

**Comment by cocoa-xu** [comment]
Not a problem. :)

Actually, after I commented about the benchmark, I found `IO.binstream`. I thought to myself, how can I miss this function? So I deleted that comment, and then I found `IO.binstream` just does not accept `BitString` either... sad

**Comment by fhunleth** [comment]
@cocoa-xu You had a change in there to mark the function as `ERL_NIF_DIRTY_JOB_IO_BOUND`. I just noticed that `spi_transfer` did not have the flag! 😱  Of all the functions, that one should have it. Could you add it and double check that it works?

Regarding binaries and splitting them, last night I was thinking of using `:binary.part/3` to split them up into a list. I got an email from GitHub that made it sound like you tried it, but now I can't find it. A few months ago, I was thinking of #52.

**Comment by cocoa-xu** [comment]
Oh sorry, I'll just put the benchmark function here!

```
defmodule ChunkTest do
    def test do
        test_sizes = for i <- 0..24, into: [], do: Bitwise.bsl(1, i)
        chunk_size = 4096
        repeat_time = 10
        time_unit = :microsecond

        for total_bytes <- test_sizes do
            bytes = for _ <- 1..total_bytes, into: "", do: <<Enum.random(0..255)>>

            start_time = :os.system_time(time_unit)
            to_u8_chunk_every_map(bytes, chunk_size)
            end_time = :os.system_time(time_unit)
            diff = Float.round((end_time - start_time) / 1_000 / repeat_time, 2)
            IO.puts("to_u8_chunk_every_map: bytes: #{total_bytes}, chunk size: #{chunk_size}, elapsed time: #{diff} ms")

            start_time = :os.system_time(time_unit)
            manually(bytes, chunk_size)
            end_time = :os.system_time(time_unit)
            diff = Float.round((end_time - start_time) / 1_000 / repeat_time, 2)
            IO.puts("manually:              bytes: #{total_bytes}, chunk size: #{chunk_size}, elapsed time: #{diff} ms")
        end
    end

    def to_u8_chunk_every_map(binary, chunk_size) do
        binary
        |> :binary.bin_to_list()
        |> Enum.chunk_every(chunk_size)
        |> Enum.map(&IO.iodata_to_binary(&1))
    end

    def manually(binary, chunk_size) do
        total_bytes = byte_size(binary)
        full_chunks = div(total_bytes, chunk_size)
        chunks = 
            if full_chunks > 0 do
                for i <- 0..(full_chunks-1), reduce: [] do
                    acc -> [:binary.part(binary, chunk_size * i, chunk_size) | acc]
                end
            else
                []  
            end
        remaining = rem(total_bytes, chunk_size)
        chunks =
            if remaining > 0 do
                [:binary.part(binary, chunk_size * full_chunks, remaining) | chunks]
            else
                chunks
            end
        Enum.reverse(chunks)
    end
end
```

I'll have a try to mark the `spi_transfer` function as `ERL_NIF_DIRTY_JOB_IO_BOUND` and test it in a second!

**Comment by fhunleth** [comment]
Great. Thanks!

**Comment by cocoa-xu** [comment]
> @cocoa-xu You had a change in there to mark the function as `ERL_NIF_DIRTY_JOB_IO_BOUND`. I just noticed that `spi_transfer` did not have the flag! 😱 Of all the functions, that one should have it. Could you add it and double check that it works?
> 
> Regarding binaries and splitting them, last night I was thinking of using `:binary.part/3` to split them up into a list. I got an email from GitHub that made it sound like you tried it, but now I can't find it. A few months ago, I was thinking of #52.

And I just found that `spi_transfer` is not marked as `ERL_NIF_DIRTY_JOB_IO_BOUND` since the initial commit. 😂
